pkg/kv/kvserver/allocator/mma/allocator_state.go

```diff
@@ -61,17 +94,267 @@ func (a *allocatorState) computeChanges() []*pendingReplicaChange {
+	// store.
+	for storeID, ss := range a.cs.stores {
+		a.changeRateLimiter.updateForRebalancePass(&ss.adjusted.enactedHistory, now)
+		sls := a.meansMemo.getStoreLoadSummary(clusterMeans, storeID, ss.loadSeqNum)
+		if (sls.sls <= overloadSlow && ss.adjusted.enactedHistory.allowLoadBasedChanges() &&
+			ss.maxFractionPending < maxFractionPendingThreshold) || sls.fd >= fdDrain {
+			sheddingStores = append(sheddingStores, sheddingStore{StoreID: storeID, storeLoadSummary: sls})
+		}
+	}
+	// We have used storeLoadSummary.sls to filter above. But when sorting, we
+	// first compare using nls. Consider the following scenario with 4 stores
+	// per node: node n1 is at 90% cpu utilization and has 4 stores each
+	// contributing 22.5% cpu utilization. Node n2 is at 60% utilization, and
+	// has 1 store contributing all 60%. When looking at rebalancing, we want to
+	// first shed load from the stores of n1, before we shed load from the store
+	// on n2.
+	slices.SortFunc(sheddingStores, func(a, b sheddingStore) int {
+		return cmp.Or(cmp.Compare(b.fd, a.fd), cmp.Compare(a.nls, b.nls), cmp.Compare(a.sls, b.sls))
+	})
+	var changes []*pendingReplicaChange
+	var disj [1]constraintsConj
+	var storesToExclude storeIDPostingList
+	var storesToExcludeForRange storeIDPostingList
+	scratchNodes := map[roachpb.NodeID]*nodeLoad{}
+	for _, store := range sheddingStores {
+		// TODO(sumeer): For remote stores that are cpu overloaded, wait for them
+		// to shed leases first. See earlier longer to do.
+
+		doneShedding := false
+		ss := a.cs.stores[store.StoreID]
+		if ss.NodeID == a.nodeID && store.storeCPUSummary <= overloadSlow {
+			// This store is local, and cpu overloaded. Shed leases first.
+			//
+			// TODO: is this path (for StoreRebalancer) also responsible for
+			// shedding leases for fdDrain and fdDead?
+			//
+			// TODO(kvoli): This path doesn't need to be responsible for shedding
+			// leases for when a store is marked as fdDrain, see
+			// `Store.SetDraining`[^*]. For fdDead, assuming liveness information is
+			// sycned with leasing (approx.), we can do nothing and the lease should
+			// have already expired. Likewise for fdDead, we can likely get away
+			// without doing anything here (rebalanceStores), but will need to add
+			// shedding functionality when replacing replicate_queue.go.
+			// [^*]: https://github.com/sumeerbhola/cockroach/blob/4f4139c2e9da8ce99cab640b41ce45b679425db8/pkg/kv/kvserver/store.go#L1868-L1868
+			//
+			// NB: any ranges at this store that don't have pending changes must
+			// have this local store as the leaseholder.
+			for _, rangeID := range ss.adjusted.topKRanges {
+				rstate := a.cs.ranges[rangeID]
+				if len(rstate.pendingChanges) > 0 {
+					// If the range has pending changes, don't make more changes.
+					continue
+				}
+				if !a.ensureAnalyzedConstraints(rstate) {
+					continue
+				}
+				cands, _ := rstate.constraints.candidatesToMoveLease()
```

Comment:
TODO: add in logic to strictly pick first preference which has >0 replicas. (move to todo in the code)

Reply:
Added todo in `candidatesToMoveLease`